deeplake/core/chunk_engine.py

```diff
@@ -714,63 +722,149 @@ def _samples_to_chunks(
+            ):
+                lengths = np.zeros(len(samples), dtype=np.uint32)
+                for i, s in enumerate(samples):
+                    try:
+                        lengths[i] = s.__len__()
+                    except AttributeError:  # None
+                        lengths[i] = 0
+                    except TypeError:  # Numpy scalar str
+                        lengths[i] = str(s).__len__()
+        extra_args = {"lengths": lengths}
         current_chunk = start_chunk
-
         updated_chunks = []
         if current_chunk is None:
-            current_chunk = self._create_new_chunk(register)
+            current_chunk = self._create_new_chunk(
+                register and start_chunk_row is not None
+            )
+            current_chunk._update_tensor_meta_length = False
             updated_chunks.append(current_chunk)
+            if extending:
+                enc_ids.append(current_chunk.id)
+        elif extending:
+            enc_ids.append(None)
         enc = self.chunk_id_encoder
         tiles = {}
-        nsamples = len(samples)
         if register and update_commit_diff:
             commit_diff = self.commit_diff
         if progressbar:
             pbar = tqdm(total=len(samples))
+        if not isinstance(samples, list) and not (
+            isinstance(samples, np.ndarray) and self._numpy_extend_optimization_enabled
+        ):
+            # Note: in the future we can get rid of this conversion of sample compressed chunks too by predicting the compression ratio.
+            samples = list(samples)
         while len(samples) > 0:
             num_samples_added = current_chunk.extend_if_has_space(
-                samples, update_tensor_meta=update_tensor_meta
+                samples, update_tensor_meta=update_tensor_meta, **extra_args
             )  # type: ignore
             if register_creds:
                 self.register_new_creds(num_samples_added, samples)
             if num_samples_added == 0:
-                current_chunk = self._create_new_chunk(register, row=start_chunk_row)
+                current_chunk = self._create_new_chunk(
+                    register and start_chunk_row is not None, row=start_chunk_row
+                )
+                current_chunk._update_tensor_meta_length = False
                 if start_chunk_row is not None:
                     start_chunk_row += 1
+                elif register:
+                    enc_ids.append(current_chunk.id)
+                    enc_count.append(0)
                 updated_chunks.append(current_chunk)
             elif num_samples_added == PARTIAL_NUM_SAMPLES:
                 sample = samples[0]
-                if register and sample.is_first_write:
-                    enc.register_samples(1)
-                if sample.is_last_write:
+                if sample.is_first_write:
                     if register:
-                        self.tile_encoder.register_sample(sample, self.num_samples - 1)
```

Comment:
Dont register to tile encoder yet. We have to update encoders, version control and tensor meta all at once to minimize risk of data corruption.